### PR TITLE
Remove pbench-generate-token agent CLI functionality

### PR DESCRIPTION
**Pull request by npalaska**
- Since we moved away from the internal user management in Pbench-server, we can no longer use username and password to create new tokens.
- The new functionality to generate long-lasting tokens will be available via a new server API calls and the dashboard. PR #3368 

PBENCH-948

### CODE REVIEW

File: contrib/containerized-pbench/pbench_demo
Comment (dbutenhof) [inline contrib/containerized-pbench/pbench_demo:37]: The comment is incorrect, since you dereferenced the `token_location` above. However, the code is also wrong: I think you meant `--token=${token}`.

I also kinda wonder whether we want to require that the token be specified in a file, vs allowing `${1}` to be the actual token, e.g., `pbench_demo $(< /file)` if you really *want* or simply `pbench_demo <token>`. Or we could be fancy and allow `pbench_demo @<file>` as well as `pbench_demo <token>`...

**Review by dbutenhof — COMMENTED** [review]
This came up in discussion this morning, and Peter was arguing against removing `pbench-generate-token`. While the normal way to get a Keycloak token is via the redirect tango, our functional tests *are* currently getting a token via Keycloak API. How practical this is in normal use, I don't know, given the bearer token lifetimes and refresh token handling: but do we want to reconsider this again??? 😕

**Comment by npalaska** [comment]
> This came up in discussion this morning, and Peter was arguing against removing `pbench-generate-token`. While the normal way to get a Keycloak token is via the redirect tango, our functional tests _are_ currently getting a token via Keycloak API. How practical this is in normal use, I don't know, given the bearer token lifetimes and refresh token handling: but do we want to reconsider this again???

But to do that we will have to expose the Keycloak APIs so the agent can call them. Can we do that with the integration lab Keycloak instance? Also is the integration lab Keycloak instance allow the CLI-based username-password authentication flow?

**Comment by dbutenhof** [comment]
> > This came up in discussion this morning, and Peter was arguing against removing `pbench-generate-token`. While the normal way to get a Keycloak token is via the redirect tango, our functional tests _are_ currently getting a token via Keycloak API. How practical this is in normal use, I don't know, given the bearer token lifetimes and refresh token handling: but do we want to reconsider this again???
> 
> But to do that we will have to expose the Keycloak APIs so the agent can call them. Can we do that with the integration lab Keycloak instance? Also is the integration lab Keycloak instance allow the CLI-based username-password authentication flow?

You're the Keycloak expert here. The thread of the discussion was that the functional test integration shows it's possible. That doesn't necessarily mean it's feasible for general use, and if it depends on specific Keycloak configuration that might be a security issue, that's probably a problem.

**Comment by npalaska** [comment]
> You're the Keycloak expert here. The thread of the discussion was that the functional test integration shows it's possible. That doesn't necessarily mean it's feasible for general use, and if it depends on specific Keycloak configuration that might be a security issue, that's probably a problem.

Yeah I mean it really depends on how the Keycloak instance in the integration lab is going to get configured. If we decide for security purposes that no one can login using the username/password through CLI (disallowing the password grant-type) then no we can't use the CLI to generate tokens. However, that brings another point, we have discussed that we (and also the integration lab instance) won't provide user registration which means all the user authentication has to happen through some identity providers. In this case, we anyway can not use the username/password on the CLI to generate the token (because our Keycloak broker is not an identity provider). Right? What we do for the functional test is just a workaround because 1. we can not use the dashboard in the functional test 2. Functional test user identity is created using some identity provider, it was created by the broker.

File: contrib/containerized-pbench/pbench_demo
Comment (webbnh) [inline contrib/containerized-pbench/pbench_demo:23]: Rather than "authentication token", I think we're calling it an "API Key".

File: contrib/containerized-pbench/pbench_demo
Comment (webbnh) [reply contrib/containerized-pbench/pbench_demo:37]: Under the assumption that this PR is _not_ going to modify `pbench-results-move` (and given that  the tool expects the value specified by `--token` to be the token string), I would just have the `pbench_demo` script require that the token be available in an environment variable which it references here and let the user worry about how to get set that up.

(Yeah, it would be cool if `pbench-results-move` supported the `@<file>` syntax, but that's for a different PR.)

File: server/pbenchinacan/load-canned-data.sh
Comment (webbnh) [inline server/pbenchinacan/load-canned-data.sh:36]: Given that `token_location` is not used except for the line immediately following its definition, let's skip defining it.

Also, given that the token string is a required parameter and that script will behave in weird ways if it's omitted, you should add a check which exits with an error if it's not provided.

**Review by webbnh — CHANGES_REQUESTED** [review]
> But to do that we will have to expose the Keycloak APIs so the agent can call them.

At least in theory, we can write a CLI tool which does anything which the Dashboard does.  (It might not be pretty or easy, but it should be just a "simple matter of programming".)  In order to be able to push a result, the Agent already has to know where the Pbench Server is, so it can just ask the Pbench Server where the Keycloak service is (so that we don't have to "expose" the Keycloak any more than that).

> Also is the integration lab Keycloak instance allow the CLI-based username-password authentication flow?

Presumably, the new `pbench-token-generate` would call the Pbench Server, then call the Keycloak broker, then handle the redirect to the OIDC identity provider, supply the username and password to it in a secure way, then ride the redirect back and obtain the access token and the refresh token, and finally hit the Pbench Server for an API key which it would return.  (Or, perhaps it can return the access token, if the user can get the job done before it expires.)

I'm not thrilled by this model, because I think that API keys will be created rarely -- but used repeately -- and that we won't need a CLI tool to do it.  (I.e, the user will create their key _once_ and then copy it to wherever they need it, so it won't matter that there is no CLI to create it -- using the Dashboard will be sufficient, and maybe even _better_, because they won't need to install the Agent in the place where they do the one-time key creation.

> it really depends on how the Keycloak instance in the integration lab is going to get configured.

It shouldn't...`pbench-token-generate` shouldn't be dependent upon any particular Pbench deployment configuration....

> What we do for the functional test is just a workaround because
> 1. we can not use the dashboard in the functional test
> 2. Functional test user identity is not created using some identity provider, it was created by the broker.

I wouldn't call what we do in the functional test a "workaround" so much as an "alternate solution".  We couldn't use the Dashboard because the functionality that we require isn't there.  Once we have API keys, I expect that we'll switch most of the functional testing to use those, and we'll get away from authentication altogether (and just concentrate on authorization...).  OIDC has freed us from the whole credentials biz, even in the Dashboard, thanks to all the redirection stuff, so it's just not a concern for us, now.

All that said, I got a couple of small requests for this PR, below.

**Comment by npalaska** [comment]
@webbnh I still do not understand how the user would be able to use the CLI to login when the identity provider is not the broker, it's either Redhat SSO or Google or something else. How to achieve these redirects without the presence of a browser and then securely insert user credentials on the identity provider login page? I guess theoretically, we can do it with curls or Python-requests but wouldn't it be messy?

**Comment by npalaska** [comment]
> Once we have API keys, I expect that we'll switch most of the functional testing to use those, and we'll get away from authentication altogether (and just concentrate on authorization...)

I doubt the path is gonna be that straight-forwards, with every functional test we create a new Keycloak container and a new database container. We create a brand new `tester` user and if we are storing any API keys they are all flushed before starting the functional tests. Unless we use some sort of a persistent storage volume for the API keys we'll have to do some experiments with this to see whether API key validation would work.

**Comment by dbutenhof** [comment]
> Once we have API keys, I expect that we'll switch most of the functional testing to use those, and we'll get away from authentication altogether (and just concentrate on authorization...).

Now with that I disagree strongly. Maybe, if API Key and Keycloak auth token validation were 100% identical, we could get away with that. They're not, and we need to test both. So, we'll need to *add* API Key testing (whether a full duplicate test set or a few selected cases, including of course the `POST /api/v1/key` itself) but we absolutely need to continue registering a Keycloak user and logging in.

That's not an issue, since I expect that our functional test environment will always be based off a local transient Keycloak ID provider that we completely control.

**Comment by npalaska** [comment]
Moving it to draft until we decide the final strategy on this topic.

**Comment by webbnh** [comment]
> Moving it to draft until we decide the final strategy on this topic.

I don't think the lack of a _final_ strategy needs to gate this PR.

At the most trivial, _if_ we were to continue to provide a CLI command which fetches an access token or an API key, I don't think it will be called `pbench-generate-token` (i.e., it will probably be called something more like `pbench-token-generate`), so we will be removing `lib/pbench/cli/agent/commands/generate_token.py` either way.

The paramount concern for this PR is that all the tests continue to pass.  However, once you fix `pbench/test/unit/agent/task/test_generate_token.py`, then I _think_ that will be the case.

So, I don't know of any reason why this PR cannot move forward.  (It's true that we will need to devise a method for testing each of access tokens and API keys, but I don't think we have to solve that in _this_ PR.)

**Comment by npalaska** [comment]
> So, I don't know of any reason why this PR cannot move forward. (It's true that we will need to devise a method for testing each of access tokens and API keys, but I don't think we have to solve that in _this_ PR.)

Right, I just wanted some preliminary understanding of our thinking on how to move forward. At the very least the current functionality is anyway not useful since it relies upon the internal user notion which we have to remove anyway. And the new approach is drastically different and may need some experimentation before we implement it.

**Review by dbutenhof — APPROVED** [review]
It looks odd that you fixed `load-canned-data.sh` which won't work anyway; but that's OK for now. 😆

**Comment by npalaska** [comment]
> It looks odd that you fixed `load-canned-data.sh` which won't work anyway; but that's OK for now.

Yeah, I think we should just drop this `load-canned-data.sh` file, I don't think anyone in the team is actually using it, right?

**Comment by webbnh** [comment]
> I think we should just drop this `load-canned-data.sh` file

Yes, I think it's time has come and gone.

File: server/pbenchinacan/load-canned-data.sh
Comment (webbnh) [inline server/pbenchinacan/load-canned-data.sh:1]: As we said, it's probably time to retire this script.

**Review by webbnh — COMMENTED** [review]
Looks good to go, once you fix the awkwardness in the demo script and remove `load-canned-data.sh`.

File: server/pbenchinacan/load-canned-data.sh
Comment (npalaska) [reply server/pbenchinacan/load-canned-data.sh:1]: Should we retire it as a separate commit?

Comment (webbnh) [reply server/pbenchinacan/load-canned-data.sh:1]: No:  you're already modifying it in this PR...just go ahead and make the "ultimate modification".

Comment (npalaska) [reply server/pbenchinacan/load-canned-data.sh:1]: Yeah, I was wondering if I should use 2 separate commits under this PR.

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
👍

File: server/pbenchinacan/load-canned-data.sh
Comment (webbnh) [reply server/pbenchinacan/load-canned-data.sh:1]: Once people have started reviewing a PR, you should only **add** commits to it; once it's time to merge a PR, how many commits you merge on to the main branch is a separate question.

Ideally, a PR makes "one change", and so all the commits from the development branch should be squashed into one commit on the main branch.  There are exceptions to every rule, but it _should_ be the case that every commit which ends up on the main branch has been individually and successfully tested...which is non-trivial to do in the "middle" of a PR.  So, if you feel like you want to merge more than one commit, it's probably best to split the PR.

However, "one commit" is often sub-optimal for review purposes.  So, feel free to present your development branch for review with as many commits on it as you think is useful...so long as you're comfortable with them all being squashed when it's merged.

**Pull request merged**
